test/e2e/framework/kcp.go

```diff
@@ -0,0 +1,274 @@
+// does not remain so.
+func (c *kcpServer) Ready() error {
+	if err := c.loadCfg(); err != nil {
+		return err
+	}
+	if c.ctx.Err() != nil {
+		// cancelling the context will preempt derivative calls but not this
+		// main Ready() body, so we check before continuing that we are live
+		return fmt.Errorf("failed to wait for readiness: %w", c.ctx.Err())
+	}
+	cfg := c.Config()
+	if cfg.NegotiatedSerializer == nil {
+		cfg.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	}
+	client, err := rest.UnversionedRESTClientFor(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create unversioned client: %w", err)
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	for _, endpoint := range []string{"/livez", "/readyz"} {
+		go func(endpoint string) {
+			defer wg.Done()
+			c.waitForEndpoint(client, endpoint)
+		}(endpoint)
+	}
+	wg.Wait()
+
+	for _, endpoint := range []string{"/livez", "/readyz"} {
+		go func(endpoint string) {
+			c.monitorEndpoint(client, endpoint)
+		}(endpoint)
+	}
+	return nil
+}
+
+func (c *kcpServer) loadCfg() error {
+	var loadError error
+	loadCtx, cancel := context.WithTimeout(c.ctx, 15*time.Second)
+	wait.UntilWithContext(loadCtx, func(ctx context.Context) {
+		kubeconfigPath := filepath.Join(c.artifactDir, "admin.kubeconfig")
+		if _, err := os.Stat(kubeconfigPath); os.IsNotExist(err) {
+			return // try again
+		} else if err != nil {
+			loadError = fmt.Errorf("failed to read admin kubeconfig after kcp start: %w", err)
+			return
+		}
+		rawConfig, err := clientcmd.LoadFromFile(kubeconfigPath)
+		if err != nil {
+			loadError = fmt.Errorf("failed to load admin kubeconfig: %w", err)
+			return
+		}
+		config, err := clientcmd.NewNonInteractiveClientConfig(*rawConfig, "admin", nil, nil).ClientConfig()
+		if err != nil {
+			loadError = fmt.Errorf("failed to load cluster-less client config: %w", err)
+			return
+		}
+		c.lock.Lock()
+		c.cfg = config
+		c.lock.Unlock()
+		cancel()
+	}, 100*time.Millisecond)
+	return loadError
+}
+
+func (c *kcpServer) waitForEndpoint(client *rest.RESTClient, endpoint string) {
+	var lastMsg string
+	var succeeded bool
+	loadCtx, cancel := context.WithTimeout(c.ctx, 15*time.Second)
+	wait.UntilWithContext(loadCtx, func(ctx context.Context) {
+		_, err := rest.NewRequest(client).RequestURI(endpoint).Do(ctx).Raw()
+		if err == nil {
+			c.t.Logf("success contacting %s", endpoint)
+			cancel()
+			succeeded = true
+		} else {
+			lastMsg = fmt.Sprintf("error contacting %s: %v", endpoint, err)
+		}
+	}, 100*time.Millisecond)
+	if !succeeded {
+		c.t.Errorf(lastMsg)
+	}
+}
+
+func (c *kcpServer) monitorEndpoint(client *rest.RESTClient, endpoint string) {
+	// we need a shorter deadline than the server, or else:
+	// timeout.go:135] post-timeout activity - time-elapsed: 23.784917ms, GET "/livez" result: Header called after Handler finished
+	ctx := c.ctx
+	if deadline, ok := c.t.Deadline(); ok {
+		deadlinedCtx, deadlinedCancel := context.WithDeadline(c.ctx, deadline.Add(-20*time.Second))
```

Comment:
What happens if the test's deadline is < 20 seconds from now when this line is called - won't deadlinedCtx immediately be canceled?

Reply:
We'd have to intentionally do this to ourselves, since the default timeout is 10 minutes, so that would be silly. Our simplest e2e already need to wait for the server to be ready and do some HTTP calls, so we're getting close to 20s as it is for normal runtime. I can add a guard in the code.